### DISCUSSION TITLE
OIDC /api/oidc/token endpoint performance

### DISCUSSION THREAD

**Discussion by LenardHess** [authelia/authelia#5731] [Q&A] [answered]
I've been setting up OIDC for Bookstack + Authelia v4.37.5 running on a Raspberry Pi 4 Model B rev 1.2 , but ran into the issue of Bookstack timing out after 3s before Authelia responds to the `POST /api/oidc/token` request. The logs indicate a total processing time of roughly 4-5s.

Question summary (Details below):
- Can the time be improved via tuning of any crypto operations?
- Are the (trace & debug) logs buffered?
- Can Authelia report request handling times?
- Can Authelia report millisecond timestamps in the logs?

<details>
<summary>The authelia log</summary>
```
time="2023-07-30T12:27:48+02:00" level=trace msg="Request hit" method=POST path=/api/oidc/token remote_ip=redacted
time="2023-07-30T12:27:52+02:00" level=debug msg="Access Request with id 'redacted' on client with id 'bookstack' is being processed" method=POST path=/api/oidc/token remote_ip=redacted
time="2023-07-30T12:27:52+02:00" level=trace msg="Access Request with id 'redacted' on client with id 'bookstack' response is being generated for session with type '*model.OpenIDSession'" method=POST path=/api/oidc/token remote_ip=redacted
    /* snip */
time="2023-07-30T12:27:53+02:00" level=debug msg="Access Request with id 'redacted' on client with id 'bookstack' has successfully been processed" method=POST path=/api/oidc/token remote_ip=redacted
    /* snip */
time="2023-07-30T12:27:53+02:00" level=trace msg="Replied (status=200)" method=POST path=/api/oidc/token remote_ip=redacted
```
</details>

The logs show a ~4s delay between the request being received (`"Request hit"`) and the request being handled (`"Access request ... is being processed"`. The subsequent `"has successfully been processed"` doesn't take long.  Is the second log message buffered? From a glance at the code, AuthenticateClient runs before the "is being processed" message.

<details>
<summary>My Notes on the code up to `"Access request is being processed"`</summary>
https://github.com/authelia/authelia/blob/be74b21ca62134a8d17703c7103e75f8075340a6/internal/server/handlers.go#L351
r.POST(
    oidc.EndpointPathToken,
    policyCORSToken.Middleware(
        bridgeOIDC(
            middlewares.NewHTTPToAutheliaHandlerAdaptor(
                handlers.OpenIDConnectTokenPOST
            )
        )
    )
)

List of middlewares chained together:
policyCORSToken.Middleware https://github.com/authelia/authelia/blob/be74b21ca62134a8d17703c7103e75f8075340a6/internal/middlewares/cors.go#L212
bridgeOIDC: https://github.com/authelia/authelia/blob/be74b21ca62134a8d17703c7103e75f8075340a6/internal/server/handlers.go#L297
    SecurityHeaders  https://github.com/authelia/authelia/blob/be74b21ca62134a8d17703c7103e75f8075340a6/internal/middlewares/headers.go#L8
    SecurityHeadersCSPNoneOpenIDConnect  https://github.com/authelia/authelia/blob/be74b21ca62134a8d17703c7103e75f8075340a6/internal/middlewares/headers.go#L31
    SecurityHeadersNoStore  https://github.com/authelia/authelia/blob/be74b21ca62134a8d17703c7103e75f8075340a6/internal/middlewares/headers.go#L46
    NewHTTPToAutheliaHandlerAdaptor  https://github.com/authelia/authelia/blob/be74b21ca62134a8d17703c7103e75f8075340a6/internal/middlewares/http_to_authelia_handler_adaptor.go#L72
OpenIDConnectTokenPOST  https://github.com/authelia/authelia/blob/master/internal/handlers/handler_oidc_token.go#L15
    NewAccessRequest  https://github.com/ory/fosite/blob/8098e48b2e0935855b61f3d7f924596620fdfc82/access_request_handler.go#L42
</details>

From a first glance most things up to `"Access request is being processed"` seem to be computationally cheap (setting values mostly). The one thing that sounds to me like it could be slow would be the [AuthenticateClient call](https://github.com/ory/fosite/blob/8098e48b2e0935855b61f3d7f924596620fdfc82/access_request_handler.go#L69C37-L69C37), but I didn't look how much work that call does. If it is expensive, wouldn't it be better to have the trace message before the operation? And what crypto parameters would dictate the cost of this operation?

**Comment by LenardHess** [comment]
Note - Bookstack doesn't support changing it's timeout duration as of now:
https://github.com/BookStackApp/BookStack/issues/4397

**Comment by james-d-elliott** [comment, marked as answer]
The performance should almost entirely based on the client secret hash which you can tune when you generate it, or at the current time (but unlikely in the future) you can use a plaintext secret. It could be possible that there is a time issue when obtaining the data from the database but that should have an error in our logs, I would have thought.

As to them supporting it he seems open to it if there is a clear need, this is a reasonable and clear need as the [specifications](https://datatracker.ietf.org/doc/html/rfc6819#section-5.1.4.1.3) themselves urge providers to not store the secret in clear text without a need (specifically things like `client_secret_jwt` client auth). 

To be fair the other method to store them would be encrypted however hashed (or hashed PLUS encrypted i.e. salt and pepper) is clearly superior, we will likely adopt a hashed only except for `client_secret_jwt` enforcement and store them encrypted in those such cases. i.e. in line with the spec.

**Reply by LenardHess** [reply]
Doing a measurement for the hash, I'm getting 4.24s for the underlying hashing operation - that seems to have been the culprit, thanks!

```sh
$ time authelia crypto hash generate pbkdf2 --variant sha512 --iterations 310000 --password 123
Digest: $pbkdf2-sha512$310000$ZEI1/Th26WqWe2Q9FugcNA$jNMFR/0tOE9JCFlthw2JmaqxDohJtxq3pWWnPMP3o1KBZysVprB4xAfF2vmpu9m8kMrreMCAiCP32acUjaXIIQ
real    0m 4.24s
user    0m 4.16s
sys     0m 0.08s
```

**Reply by james-d-elliott** [reply]
Yeah that'll be it, you can obviously tune those iterations or the algorithm to something suitable for your needs.

**Reply by LenardHess** [reply]
I'll be adding some hints to the bookstack docs for this specific issue in a pull request shortly.

Does authelia have any metrics for the time taken by endpoints?

**Reply by james-d-elliott** [reply]
I'd rather the general information is updated to include a note that the token endpoint time is affected based on the secret hash choices, and that may affect some clients if they have a short timeout. Otherwise it's not very green.

We don't specifically track individual endpoints, we do track request times and authentication times however. We could add one for this.

**Reply by LenardHess** [reply]
Ok, i'll update the docs pull request to something more generic.

Without generic endpoint time duration - For this specific endpoint it might make sense to either move the "Access request ... is being processed" trace msg before the client secret validation or add a separate debug noting that the client secret is being checked.
